### ISSUE TITLE
Matrix corresponding to convolution by a 2D kernel (convmtx2)

### ISSUE THREAD

**Issue by carlosgmartin** [pytorch/pytorch#26781]
## 🚀 Feature
A function `convmtx2` such that

    convmtx2(kernel, image.shape[1:]) @ image.flatten() == torch.nn.functional.conv2d(image[None], kernel)[0].flatten()

for any tensors `image` of shape `(in_channels, image_height, image_width)` and `kernel` of shape `(out_channels, in_channels, kernel_height, kernel_width)`.

## Motivation

Like its [MATLAB counterpart](convmtx2), `convmtx2` computes the matrix corresponding to convolution by a 2D kernel. This matrix is a doubly block [Toeplitz matrix](https://en.wikipedia.org/wiki/Toeplitz_matrix#Discrete_convolution).

This feature has been requested elsewhere. See [here](https://discuss.pytorch.org/t/obtaining-toeplitz-matrix-for-convolution/52968/3) and [here](https://stackoverflow.com/questions/56702873/is-there-an-function-in-pytorch-for-converting-convolutions-to-fully-connected-n) for example.

**Comment by carlosgmartin** [comment]
Here is a slow "solution" that works by convolving with all possible one-hot images:

```
import torch
def test(convmtx2):
    out_channels = 5
    in_channels = 3
    image_height = 10
    image_width = 20
    kernel_height = 4
    kernel_width = 6
    while True:
        image = torch.rand((in_channels, image_height, image_width))
        kernel = torch.rand((out_channels, in_channels, kernel_height, kernel_width))
        a = convmtx2(kernel, image.shape) @ image.flatten()
        b = torch.nn.functional.conv2d(image[None], kernel)[0].flatten()
        print((a - b).abs().max())

def convmtx2_slow(kernel, image_shape):
    assert image_shape[0] == kernel.shape[1]
    columns = []
    for c in range(kernel.shape[1]):
        for i in range(image_shape[1]):
            for j in range(image_shape[2]):
                onehot = torch.zeros(image_shape)
                onehot[c, i, j] = 1
                column = torch.nn.functional.conv2d(onehot[None], kernel)[0].flatten()
                columns.append(column)
    return torch.stack(columns, 1)

test(convmtx2_slow)
```

**Comment by carlosgmartin** [comment]
More generally, to take into account things like stride, dilation, biases, etc., it would be nice to have a function that simply converts a `torch.nn.Conv2d` to its corresponding `torch.nn.Linear`. These are all just instances of matrix multiplication with the kernel, image, or their convolution: https://datascience.stackexchange.com/a/44349/52298.

**Comment by carlosgmartin** [comment]
Here's an implementation that *almost* works for an arbitrary number of dimensions:

```
def convmatrix(kernel, image_shape):
    # kernel: (out_channels, in_channels, kernel_height, kernel_width, ...)
    # image: (in_channels, image_height, image_width, ...)
    assert image_shape[0] == kernel.shape[1]
    assert len(image.shape[1:]) == len(kernel.shape[2:])

    result_dims = torch.tensor(image_shape[1:]) - torch.tensor(kernel.shape[2:]) + 1

    m = torch.zeros((
        kernel.shape[0],
        *result_dims,
        *image_shape
    ))

    for index in itertools.product(*map(range, result_dims)):
        m[
            :,
            index[0],
            index[1],
            :,
            index[0]:kernel.shape[2+0]+index[0],
            index[1]:kernel.shape[2+1]+index[1]
        ] = kernel

    return m.flatten(end_dim=len(image.shape[1:])).flatten(start_dim=1)
```

I'm trying to find a way to vectorize the slices and avoid the loop (while at the same time making it dimension-independent).

**Comment by carlosgmartin** [comment]
The following seems to work for an arbitrary number of dimensions:

```
def convmatrix(kernel, image_shape):
    result_dims = torch.tensor(image_shape[1:]) - torch.tensor(kernel.shape[2:]) + 1

    d = len(kernel.shape[2:])

    x = torch.nn.functional.pad(kernel, tuple(torch.stack((torch.zeros_like(result_dims),result_dims.flip(0)),1).flatten()))
    grid = torch.stack(torch.meshgrid(*map(torch.arange, (
        kernel.shape[0],
        *result_dims,
        *image_shape
    ))), -1)
    y = ((grid[..., d+2:] - grid[..., 1:d+1]) % torch.tensor(image_shape[1:]))
    m = x[(
        grid[..., 0],
        grid[..., d+1],
        *y.permute(*torch.arange(len(y.shape)).roll(1,0)),
    )]
    return m.flatten(end_dim=len(kernel.shape[2:])).flatten(start_dim=1)

if __name__ == '__main__':

    if True: # test 1d
        out_channels = 2
        in_channels = 3
        image_height = 10
        kernel_height = 4

        image = torch.rand((in_channels, image_height))
        kernel = torch.rand((out_channels, in_channels, kernel_height))
        a = torch.nn.functional.conv1d(image[None], kernel)[0].flatten()
        b = convmatrix(kernel, image.shape) @ image.flatten()
        print((a - b).abs().max())

    if True: # test 2d
        out_channels = 2
        in_channels = 3
        image_height = 20
        kernel_height = 5
        image_width = 30
        kernel_width = 10

        image = torch.rand((in_channels, image_height, image_width))
        kernel = torch.rand((out_channels, in_channels, kernel_height, kernel_width))
        a = torch.nn.functional.conv2d(image[None], kernel)[0].flatten()
        b = convmatrix(kernel, image.shape) @ image.flatten()
        print((a - b).abs().max())
```

The code seems a bit ugly though. Perhaps it can be cleaned up somehow.

**Comment by carlosgmartin** [comment]
Here's a version that works for 2D specifically and seems faster:

```
def convmatrix2d(kernel, image_shape):
    # kernel: (out_channels, in_channels, kernel_height, kernel_width, ...)
    # image: (in_channels, image_height, image_width, ...)
    assert image_shape[0] == kernel.shape[1]
    assert len(image_shape[1:]) == len(kernel.shape[2:])
    result_dims = torch.tensor(image_shape[1:]) - torch.tensor(kernel.shape[2:]) + 1
    m = torch.zeros((
        kernel.shape[0], 
        *result_dims, 
        *image_shape
    ))
    for i in range(m.shape[1]):
        for j in range(m.shape[2]):
            m[:,i,j,:,i:i+kernel.shape[2],j:j+kernel.shape[3]] = kernel
    return m.flatten(0, len(kernel.shape[2:])).flatten(1)
```

It would be nice if the two loops could be vectorized somehow.

**Comment by carlosgmartin** [comment]
As an aside does anyone know how to properly `@torch.jit.script` the above function?

**Comment by Mikkolehtimaki** [comment]
Is there a neat way to implement stride > 1? Just deleting every other row from the resulting matrix will lead into wrong results.

**Comment by samuelemarro** [comment]
@Mikkolehtimaki Did you by chance find a solution? I'm struggling to implement it too.

**Comment by samuelemarro** [comment]
For future reference, a quick and dirty way to implement the generalized conversion is to apply the convolution to a (reshaped) identity matrix:
````
def to_matrix(conv, image_shape, output_shape):
    identity = torch.eye(np.prod(image_shape).item()).reshape([-1] + list(image_shape))
    output = F.conv2d(identity, conv.weight, None, conv.stride, conv.padding)
    W = output.reshape(-1, np.prod(output_shape).item()).T
    b = torch.stack([torch.ones(output_shape[1:]) * bi for bi in conv.bias])
    b = b.reshape(-1, np.prod(output_shape).item())
    return W, b
````
I honestly don't know why W must be transposed, but it's the only way to make it work.

**Comment by Mikkolehtimaki** [comment]
> @Mikkolehtimaki Did you by chance find a solution? I'm struggling to implement it too.

Unfortunately I did not, I had to move on.

**Comment by adam2392** [comment]
Has there been any progress on this? I would definitely like to have access to the underlying toeplitz matrix that performs the 2D convolution for a given kernel and image shape.

Ideally the solution is vectorized and doesn't require a double for loop to set it up too :p.

**Comment by Mikkolehtimaki** [comment]
I did add support for zero-padding. Turned out to be quite simple. I ended up not needing strides so those are still missing.

```
def convmatrix2d(kernel, image_shape, padding: int=0):
    """
    kernel: (out_channels, in_channels, kernel_height, kernel_width, ...)
    image: (in_channels, image_height, image_width, ...)
    padding: assumes the image is padded with ZEROS of the given amount
    in every 2D dimension equally. The actual image is given with unpadded dimension.
    """

    # If we want to pad, request a bigger matrix as the kernel will convolve
    # over a bigger image.
    if padding:
        old_shape = image_shape
        pads = (padding, padding, padding, padding)
        image_shape = (image_shape[0], image_shape[1] + padding*2, image_shape[2]
                       + padding*2)
    else:
        image_shape = tuple(image_shape)
    assert image_shape[0] == kernel.shape[1]
    assert len(image_shape[1:]) == len(kernel.shape[2:])
    assert stride == 1

    kernel = kernel.to('cpu') # always perform the below work on cpu

    result_dims = (torch.tensor(image_shape[1:]) -
                   torch.tensor(kernel.shape[2:]))//stride + 1
    mat = torch.zeros((kernel.shape[0], *result_dims, *image_shape))
    for i in range(mat.shape[1]):
        for j in range(mat.shape[2]):
            mat[:,i,j,:,i:i+kernel.shape[2],j:j+kernel.shape[3]] = kernel
    mat = mat.flatten(0, len(kernel.shape[2:])).flatten(1)

    # Handle zero padding. Effectively, the zeros from padding do not
    # contribute to convolution output as the product at those elements is zero.
    # Hence the columns of the conv mat that are at the indices where the
    # padded flattened image would have zeros can be ignored. The number of
    # rows on the other hand must not be altered (with padding the output must
    # be larger than without). So..

    # Begin by obtaining the indices where zero padded flattened image has
    # zeros from padding. These are the first and last columns and
    # rows, which in flattened form are the (width + 2) first and last
    # elements, and additionally two elements every width apart. This sequence
    # repeats for every channel in the input image.

    # We'll handle this the easy way and create a mask that accomplishes the
    # indexing
    if padding:
        mask = torch.nn.functional.pad(torch.ones(old_shape), pads).flatten()
        mask = mask.bool()
        mat = mat[:, mask]

    return mat

```

**Comment by adam2392** [comment]
Oh awesome @Mikkolehtimaki ! Unfortunately we still have that double for loop, but this is really helpful to get a MVP started for me.

Just so I can understand correctly, the matrix returned is: `mat` (d', d), where d is the dimension of your vectorized image and d' is the data after applying convolution, so to apply to your data, `X` (n, d), you would do:

```
X.dot(mat.T)
```
which would give you the resulting dataset convolved with your corresponding kernel. Is the transpose in their correct?

Thanks!

**Comment by Mikkolehtimaki** [comment]
@adam2392 I don't think you can use (pytorch's) `dot` because dimensions will be inconsistent. You can set `mat` as the weight for a linear layer and feed flattened images. Or do the same manually `image.flatten(start_dim=1).mm(convmat.T)` if image has shape `[batch, channel, height, width]`, so in that sense yes the transpose is correct if you want to maintain batch as first dimension.

**Comment by vadimkantorov** [comment]
This could be useful for experimenting with some sort of dynamic convolution kernels that are space-variant (or some nice way of playing with local attention)

**Comment by adam2392** [comment]
Yeah would be great to have this functionality.

**Comment by vadimkantorov** [comment]
@cpuhrsch this could probably also used for experimenting with implementing convolutions for nestedtensor inputs without temporary conversion to padded format
